src/kpro_lib.erl

```diff
@@ -0,0 +1,301 @@
+%%       only exceptions and process links are propagated back to caller
+%%       other side-effects like monitor references are not handled.
+ok_pipe(FunList, Timeout) ->
+  with_timeout(fun() -> do_ok_pipe(FunList) end, Timeout).
+
+%% @doc Same as `ok_pipe/2' with `infinity' as default timeout.
+ok_pipe(FunList) ->
+  ok_pipe(FunList, infinity).
+
+%% @doc Parse comma separated endpoints in a string into a list of
+%% `{Host::string(), Port::integer()}' pairs.
+%% Endpoints may or may not start with protocol prefix (non case sensitive):
+%% `PLAINTEXT://', `SSL://', `SASL_PLAINTEXT://' or `SASL_SSL://'.
+%% The first arg is to filter desired endpoints from parse result.
+-spec parse_endpoints(kpro:protocol() | undefined, string()) ->
+        [kpro:endpoint()].
+parse_endpoints(Protocol, Str) ->
+  Eps0 = string:tokens(Str, ",\n"),
+  L = [parse_endpoint(string:to_lower(Ep)) || Ep <- Eps0, Ep =/= ""],
+  [EP || {P, EP} <- L, Protocol =:= P].
+
+%% @doc Return number of bytes in the given `iodata()'.
+-spec data_size(iodata()) -> count().
+data_size(IoData) ->
+  data_size(IoData, 0).
+
+%% @doc Encode primitives.
+-spec encode(primitive_type(), kpro:primitive()) -> iodata().
+encode(boolean, true) -> <<1:8/?INT>>;
+encode(boolean, false) -> <<0:8/?INT>>;
+encode(int8,  I) when is_integer(I) -> <<I:8/?INT>>;
+encode(int16, I) when is_integer(I) -> <<I:16/?INT>>;
+encode(int32, I) when is_integer(I) -> <<I:32/?INT>>;
+encode(int64, I) when is_integer(I) -> <<I:64/?INT>>;
+encode(varint, I) when is_integer(I) -> kpro_varint:encode(I);
+encode(nullable_string, ?null) -> <<-1:16/?INT>>;
+encode(nullable_string, Str) -> encode(string, Str);
+encode(string, Atom) when is_atom(Atom) ->
+  encode(string, atom_to_binary(Atom, utf8));
+encode(string, <<>>) -> <<0:16/?INT>>;
+encode(string, L) when is_list(L) ->
+  encode(string, iolist_to_binary(L));
+encode(string, B) when is_binary(B) ->
+  Length = size(B),
+  <<Length:16/?INT, B/binary>>;
+encode(bytes, ?null) -> <<-1:32/?INT>>;
+encode(bytes, B) when is_binary(B) orelse is_list(B) ->
+  Size = kpro_lib:data_size(B),
+  case Size =:= 0 of
+    true  -> <<-1:32/?INT>>;
+    false -> [<<Size:32/?INT>>, B]
+  end;
+encode(records, B) ->
+  encode(bytes, B).
+
+%% @doc Decode primitives.
+-spec decode(kpro:primitive_type(), binary()) -> {primitive(), binary()}.
+decode(boolean, Bin) ->
+  <<Value:8/?INT, Rest/binary>> = Bin,
+  {Value =/= 0, Rest};
+decode(int8, Bin) ->
+  <<Value:8/?INT, Rest/binary>> = Bin,
+  {Value, Rest};
+decode(int16, Bin) ->
+  <<Value:16/?INT, Rest/binary>> = Bin,
+  {Value, Rest};
+decode(int32, Bin) ->
+  <<Value:32/?INT, Rest/binary>> = Bin,
+  {Value, Rest};
+decode(int64, Bin) ->
+  <<Value:64/?INT, Rest/binary>> = Bin,
+  {Value, Rest};
+decode(varint, Bin) ->
+  kpro_varint:decode(Bin);
+decode(string, Bin) ->
+  <<Size:16/?INT, Rest/binary>> = Bin,
+  copy_bytes(Size, Rest);
+decode(bytes, Bin) ->
+  <<Size:32/?INT, Rest/binary>> = Bin,
+  copy_bytes(Size, Rest);
```

Comment:
So decoding a binary embedded binaries always copies.  Have you seen improved memory behaviour from this decision?

Reply:
I didn't do any profiling in this regard. And I'm not sure if @id did any.

Reply:
We did have a memory leak in the earlier versions of brod. I tried to nail it down by copying binaries, but in the end it turned out to be in a completely different place. I left copying as is and didn't test memory behaviour without copying binaries after the memory leak bug was fixed.

Could be worth trying.